source/fuzz/fuzzer_pass_interchange_zero_like_constants.cpp

```diff
@@ -0,0 +1,123 @@
+    // Add declaration of equivalent scalar constant
+    auto kind = constant->type()->kind();
+    if (kind == opt::analysis::Type::kBool ||
+        kind == opt::analysis::Type::kInteger ||
+        kind == opt::analysis::Type::kFloat) {
+      return FindOrCreateZeroConstant(declaration->type_id());
+    }
+  }
+
+  return 0;
+}
+
+void FuzzerPassInterchangeZeroLikeConstants::AddUseToReplace(
+    opt::Instruction* use_inst, uint32_t use_index, uint32_t replacement_id,
+    std::vector<std::pair<protobufs::IdUseDescriptor, uint32_t>>&
+        uses_to_replace) {
+  // Only consider this use if it is in a block
+  if (GetIRContext()->get_instr_block(use_inst)) {
+    // Get index of the operands with respect to just the input operands.
+    // (|use_index| refers to all the operands, not just the In operands)
+    uint32_t in_operand_index =
+        use_index - use_inst->NumOperands() + use_inst->NumInOperands();
+    auto id_use_descriptor =
+        MakeIdUseDescriptorFromUse(GetIRContext(), use_inst, in_operand_index);
+    uses_to_replace.emplace_back(
+        std::make_pair(id_use_descriptor, replacement_id));
+  }
+}
+
+void FuzzerPassInterchangeZeroLikeConstants::Apply() {
+  // Find all constants
+  auto constants = GetIRContext()->GetConstants();
+
+  // Make vector keeping track of all the uses we want to replace.
+  // This is a vector of pairs, where the first element is an id use descriptor
+  // identifying the use of a constant id and the second is the id that should
+  // be used to replace it.
+  std::vector<std::pair<protobufs::IdUseDescriptor, uint32_t>> uses_to_replace;
+
+  for (auto constant : constants) {
+    uint32_t constant_id = constant->result_id();
+    uint32_t toggled_id = FindOrCreateToggledConstant(constant);
+
+    if (!toggled_id) {
+      // Not a zero-like constant
+      continue;
+    }
+
+    // Record synonymous constants
+    ApplyTransformation(
+        TransformationRecordSynonymousConstants(constant_id, toggled_id));
+
+    // Find all the uses of the constant and, for each, probabilistically
+    // decide whether to replace it.
+    GetIRContext()->get_def_use_mgr()->ForEachUse(
+        constant_id,
+        [this, toggled_id, &uses_to_replace](opt::Instruction* use_inst,
+                                             uint32_t use_index) -> void {
+          if (GetFuzzerContext()->ChoosePercentage(
+                  GetFuzzerContext()
+                      ->GetChanceOfInterchangingZeroLikeConstants())) {
+            AddUseToReplace(use_inst, use_index, toggled_id, uses_to_replace);
+          }
+        });
+  }
+
+  // Replace the ids
+  for (auto use_to_replace : uses_to_replace) {
+    auto transformation = TransformationReplaceIdWithSynonym(
+        use_to_replace.first, use_to_replace.second);
+    if (transformation.IsApplicable(GetIRContext(),
+                                    *GetTransformationContext())) {
+      ApplyTransformation(transformation);
+    }
```

Comment:
The problem is that there are some more conditions which must be satisfied (e.g. the index of OpAccessChain must be OpConstant, cannot be OpConstantNull) - this is checked by the UseCanBeReplacedWithSynonym method of TransformationReplaceIdWithSynonym. Maybe it would be good to make this method public or make it otherwise reusable so that it can be used directly to make the check instead of calling IsApplicable?

Reply:
Changed accordingly to the chat